lib/QubitFlatfileImport.class.php

```diff
@@ -1416,28 +1395,59 @@ public static function fetchKeymapEntryBySourceAndTargetName($sourceId, $sourceN
 
     /**
      * Map a value to its corresponding term name then return the term ID
-     * corresponding to the term name.
+     * corresponding to the term name. This function will create a new term if
+     * not found in taxonomy.
      *
-     * @param string $description        description of subject (for error output)
-     * @param string $value              value that needs to be mapped to a term ID
-     * @param array  $valueToTermNameMap array mapping possible values to term names
-     * @param array  $terms              array mapping term IDs to term names
+     * @param string $taxonomyDescription description of subject (for error output)
+     * @param string $value               value that needs to be mapped to a term ID
+     * @param mixed  $culture             row culture
+     * @param array  $terms               array mapping term IDs to term names.
+     *                                    See QubitFlatfileImport::loadTermsFromTaxonomies.
+     * @param mixed  $taxonomyId          taxonomy id of term to be created
      *
      * @return int term ID
      */
-    public function translateNameToTermId($description, $value, $valueToTermNameMap, $terms)
+    public function createOrFetchTermIdFromName($taxonomyDescription, $value, $culture, &$terms, $taxonomyId)
     {
-        if (isset($valueToTermNameMap[$value]) || 0 == count($valueToTermNameMap)) {
-            $termName = (count($valueToTermNameMap)) ? $valueToTermNameMap[$value] : $value;
+        if (empty($value)) {
+            return;
+        }
+
+        $key = false;
+        if (isset($terms[$culture])) {
+            $key = csvImportBaseTask::arraySearchCaseInsensitive($value, $terms[$culture]);
+        }
 
-            if (in_array($termName, $terms)) {
-                return array_search($termName, $terms);
+        if (false === $key) {
+            // Test if term is present in DB. Retrieve taxonomy terms from term table.
+            $query = "SELECT t.id FROM term t LEFT JOIN term_i18n ti ON t.id=ti.id \r
+                WHERE t.taxonomy_id=? AND ti.culture=? AND ti.name=?";
+
+            $result = QubitPdo::fetchOne($query, [$taxonomyId, $culture, $value]);
+
+            if (false !== $result) {
+                return $result->id;
```

Comment:
In this conditional `$value` was not found in the `$terms` array, but was in the database, so I think it would be good to add the `$value` to the `$terms` array to prevent subsequent database queries for the same `$value`?

Reply:
I think this makes sense. At one point during dev, I had a call to csvImportBaseTask::refreshTaxonomyTerms() right before this return but removed it! Added a new commit adding it back in.